zetaclient/chains/solana/signer/signer.go

```diff
@@ -0,0 +1,280 @@
+package signer
+
+import (
+	"context"
+
+	"cosmossdk.io/errors"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"github.com/near/borsh-go"
+
+	"github.com/zeta-chain/zetacore/pkg/chains"
+	"github.com/zeta-chain/zetacore/pkg/coin"
+	contract "github.com/zeta-chain/zetacore/pkg/contract/solana"
+	"github.com/zeta-chain/zetacore/x/crosschain/types"
+	observertypes "github.com/zeta-chain/zetacore/x/observer/types"
+	"github.com/zeta-chain/zetacore/zetaclient/chains/base"
+	"github.com/zeta-chain/zetacore/zetaclient/chains/interfaces"
+	"github.com/zeta-chain/zetacore/zetaclient/metrics"
+	"github.com/zeta-chain/zetacore/zetaclient/outboundprocessor"
+)
+
+var _ interfaces.ChainSigner = (*Signer)(nil)
+
+// Signer deals with signing BTC transactions and implements the ChainSigner interface
+type Signer struct {
+	*base.Signer
+
+	// client is the Solana RPC client that interacts with the Solana chain
+	client interfaces.SolanaRPCClient
+
+	// solanaFeePayerKey is the private key of the fee payer account on Solana chain
+	solanaFeePayerKey solana.PrivateKey
+
+	// gatewayID is the program ID of gateway program on Solana chain
+	gatewayID solana.PublicKey
+
+	// pda is the program derived address of the gateway program
+	pda solana.PublicKey
+}
+
+// NewSigner creates a new Bitcoin signer
+func NewSigner(
+	chain chains.Chain,
+	chainParams observertypes.ChainParams,
+	solClient interfaces.SolanaRPCClient,
+	tss interfaces.TSSSigner,
+	solanaKey solana.PrivateKey,
+	ts *metrics.TelemetryServer,
+	logger base.Logger,
+) (*Signer, error) {
+	// create base signer
+	baseSigner := base.NewSigner(chain, tss, ts, logger)
+
+	// parse gateway ID and PDA
+	gatewayID, pda, err := contract.ParseGatewayIDAndPda(chainParams.GatewayAddress)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot parse gateway address %s", chainParams.GatewayAddress)
+	}
+	logger.Std.Info().Msgf("Solana fee payer address: %s", solanaKey.PublicKey())
+
+	// create solana observer
+	return &Signer{
+		Signer:            baseSigner,
+		client:            solClient,
+		solanaFeePayerKey: solanaKey,
+		gatewayID:         gatewayID,
+		pda:               pda,
+	}, nil
+}
+
+// SignMsgWithdraw signs a message for Solana gateway withdraw/withdraw_spl instruction
+func (signer *Signer) SignMsgWithdraw(
```

Comment:
I think we can move the three next methods in a `withdraw.go` file

Reply:
yeah. good idea

Reply:
https://github.com/zeta-chain/node/pull/2560/commits/ada365437e0782d6d2f6725ca87d6c53a687118d